File: Recursion/Patterns/Pattern5/Pattern5.java
package Recursion.Patterns.Pattern5;

import java.util.*;
public class Pattern5 {
    public static void main(String[] args) {
        Set<List<Integer>> ls=fun(0,7,new ArrayList<>(),new HashSet<>(),new int[]{1,2,3,4});
        System.out.println(ls);
    }
    public static Set<List<Integer>> fun(int ind,int target,List<Integer> lst, Set<List<Integer>> lst2,int[] num){
        if(target<=0){
            return lst2;
        }
        lst2.add(new ArrayList<>(lst));
        if(ind>=num.length){
            return lst2;
        }
        lst.add(num[ind]);
        fun(ind+1,target-num[ind],lst,lst2,num);
        lst.remove(lst.size()-1);
        fun(ind+1,target,lst,lst2,num);
        return lst2;
    }
}

File: Recursion/Patterns/Pattern5/Pattern5.py
def fun(num,ind,lst,saet,target):
    if target<=0:
        return saet
    saet.add(tuple(lst))
    if ind>=len(num):
        return saet
    lst.append(num[ind])
    fun(num,ind+1,lst,saet,target-num[ind])
    lst.pop()
    fun(num,ind+1,lst,saet,target)
    return saet
out=fun([1,2,3,4],0,[],set(),7)
print(out)